openapi-manager/src/validation.rs

```diff
@@ -0,0 +1,227 @@
+    }
+
+    // Perform any additional API-specific validation.
+    api.extra_validation(
+        &openapi_doc,
+        ValidationContext::new(&mut validation_context),
+    );
+
+    if !validation_context.errors.is_empty() {
+        return Err(anyhow::anyhow!(
+            "OpenAPI document validation failed:\n{}",
+            validation_context
+                .errors
+                .iter()
+                .map(|e| e.to_string())
+                .collect::<Vec<_>>()
+                .join("\n")
+        ));
+    }
+
+    Ok(ValidationResult { extra_files: validation_context.files })
+}
+
+/// Check a file against expected contents.
+fn check_file(
+    full_path: Utf8PathBuf,
+    contents: Vec<u8>,
+) -> anyhow::Result<CheckStatus> {
+    let existing_contents =
+        read_opt(&full_path).context("failed to read contents on disk")?;
+
+    match existing_contents {
+        Some(existing_contents) if existing_contents == contents => {
+            Ok(CheckStatus::Fresh)
+        }
+        Some(existing_contents) => {
+            Ok(CheckStatus::Stale(CheckStale::Modified {
+                full_path,
+                actual: existing_contents,
+                expected: contents,
+            }))
+        }
+        None => Ok(CheckStatus::Stale(CheckStale::New { expected: contents })),
+    }
+}
+
+pub fn read_opt(path: &Utf8Path) -> std::io::Result<Option<Vec<u8>>> {
+    match fs_err::read(path) {
+        Ok(contents) => Ok(Some(contents)),
+        Err(err) if err.kind() == std::io::ErrorKind::NotFound => Ok(None),
+        Err(err) => return Err(err),
+    }
+}
+
+#[derive(Debug)]
+#[must_use]
+pub(crate) enum OverwriteStatus {
+    Updated,
+    Unchanged,
+}
+
+/// Overwrite a file with new contents, if the contents are different.
+///
+/// The file is left unchanged if the contents are the same. That's to avoid
+/// mtime-based recompilations.
+pub fn overwrite_file(
+    path: &Utf8Path,
+    contents: &[u8],
+) -> anyhow::Result<OverwriteStatus> {
+    // Only overwrite the file if the contents are actually different.
+    let existing_contents =
+        read_opt(path).context("failed to read contents on disk")?;
+
+    // None means the file doesn't exist, in which case we always want to write
+    // the new contents.
+    if existing_contents.as_deref() == Some(contents) {
+        return Ok(OverwriteStatus::Unchanged);
+    }
+
+    // Make sure the parent directory exists before trying to write any files.
+    // N.B. that it's very unlikely that `parent()` would be `None` --- why are
+    // you putting your OpenAPI document in `/`? --- but we may as well not fail
+    // if that is the case...
+    if let Some(parent) = path.parent() {
+        // Call create_dir_all unconditionally: it is idempotent and doesn't
+        // error out if the parent already exists.
+        fs_err::create_dir_all(parent).with_context(|| {
+            format!("failed to create parent directory for '{}'", path)
+        })?
```

Comment:
This previously used to be:

```rust
if !parent.exists() {
    fs_err::create_dir_all(path)...
}
```

But that isn't right -- instead, it should try creating parent directories. The exists check is also unnecessary.